_includes/parse-server/best-practice.md

```diff
@@ -10,6 +10,57 @@ Protect all Parse Server endpoints using a Firewall to mitigate the risk of mali
 - Use rate-limiting rules for public endpoints, for example limit the number of requests per IP address or per user.
 - Use very restrictive rules for private endpoints; for example limit access to Parse Dashboard to your personal network.
 
+### Security-First Mindset
+
+When developing for Parse Server, consider any Cloud Function or Cloud Trigger as a potential target for a malicious attack. In general, it's best to always assume that the sender of a request cannot be trusted. Even if you have control over the client's code when shipping it, it's simple for an attacker to replicate a request to Parse Server and modify arguments, or reverse-engineer a client app and extract hard coded keys to send requests with unexpected arguments.
+
+The following cloud code is **not recommended** and is an example of poor security practise:
+
+```js
+Parse.Cloud.define('updateEmail', async (req) => {
+  const user = await new Parse.Query(Parse.User).get(req.params.id, { useMasterKey: true });
+  if (!user) {
+    throw 'This user does not exist';
+  }
+  user.set('email', req.params.email);
+  return user.save(null, { useMasterKey: true });
+});
+```
+
+With this code, an attacker can simply guess a user to sniff out a user object and replace their email, potentially exposing private user data. This could allow the attacker to then request a password reset to the new email, and take-over the account completely.
+
+```js
+while (true) {
+  let id = '';
+  const chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghiklmnopqrstuvwxyz'.split('');
+  for (let i = 0; i < 10; i++) {
+    id += chars[Math.floor(Math.random() * chars.length)];
+  }
+  try {
+    const user = await Parse.Cloud.run('updateEmail', {id});
+    console.log(`Here is a full user object: `, user.toJSON());
+    return;
+  } catch (e) {
+    console.log(`Not a valid id: ${id}`);
+  }
+}
+```
+
+Regardless of how secure your Parse Server is with ACLs and CLPs, a few lines of insecure Cloud Functions can undo it.
+
+A secure solution would be to ensure that users can only update their email, such as:
+
+```js
+Parse.Cloud.define('updateEmail', async (req) => {
+  req.user.set('email', req.params.email);
+  await req.user.save(null, { useMasterKey: true });
```

Comment:
Wouldn't it be best practice to avoid any operation using `useMasterKey: true` but instead give only the user write permission in the object ACL? This would mitigate the risk that through a developer mistake in code someone could execute operations because they are using the master key; instead any permission would be bound to the user's session token.